### PR TITLE
Update beacon state locks

### PR DESCRIPTION
**Pull request by mcdee**
This patch fixes a number of synchronisation issues in the beacon state.  The specific changes are:

  - getters in `BeaconState` have been broken out in to a public getter, which read-locks the state mutex, and a package-local getter, which doesn't.
  - changes to `BeaconState.sharedFieldReferences` are protected by a write lock rather than a read lock, to avoid concurrent map access/write
  - `BeaconState.Copy()` obtains a single write lock and then avoids calling any functions that fetch a read lock

Obviously any changes to locks needs to be carefully examined.  I have run this locally for a day or so, but well worth someone going over this one with a fine-toothed comb.

Incidentally, the package I used to help track down the issues fixed in the PR was https://github.com/sasha-s/go-deadlock.  It is a drop-in replacement for `sync`'s mutexes that tracks usage and highlights potential deadlock situations.

One thing worth noting: @nisdas raised the question of why deadlocks are being seen now when the state code has hardly changed in months.  I believe this is down to two changes: first, init sync is a lot faster than it used to be, and as such there is a lot more write activity in a given time period.  Second, there is more monitoring software coming on-stream, so more reads against the beacon state.

### CODE REVIEW

**Comment by nisdas** [comment]
@mcdee would it be possible to gate this behind a flag ? I would like to run it for a while in our internal deployment before we make this the default for everyone.

File: beacon-chain/state/getters.go [hunk 1]
```diff
@@ -106,6 +143,34 @@ func (b *BeaconState) CloneInnerState() *pbp2p.BeaconState {
 	if b == nil || b.state == nil {
 		return nil
 	}
+	b.lock.RLock()
```
Comment (prestonvanloon) [inline beacon-chain/state/getters.go:146]: Isn't this going to cause a deadlock if you're not using NewBeaconStateLocks?

**Comment by rauljordan** [comment]
Hi @mcdee, please resolve conflicts and we can review right after

**Comment by prestonvanloon** [comment]
Moved some commentary to docs.go so it would appear in the godoc "overview" section. 5fc352d0436bffd1125b131326e7df670aad5a78

**Pull request merged**